runtime/lua/nvim/meta.lua

```diff
@@ -0,0 +1,51 @@
+
+local meta = {}
+
+---
+-- Provide a "strongly typed" dictionary in Lua.
+--
+-- Does not allow insertion or deletion after creation.
+-- Only allows retrieval of created keys are allowed.
+meta.Enum = {
+  new = function(self, map)
+    return setmetatable(map, self)
+  end,
+
+  __index = function(t, k)
+    error("attempt to get unknown enum " .. k .. "from " .. tostring(t), 2)
+  end,
+
+  __newindex = function(t, k, v)
+    error(
+      string.format("attempt to update enum table with %s, %s, %s", t, k, v),
+      2)
+  end
+}
+
+---
+-- Provide a dictionary that will continue providing empty dictionaries upon access.
+--
+-- This allows you to do something like:
+--  local myEmpty =- meta.EmptyDictionary({a = 'b'})
+--  if myEmpty.b.a.c.d.e.f.g.i == nil then
+--      // Do some error stuff here
+--  end
+meta.EmptyDictionary = {
```

Comment:
In python this is called `defaultdict`. It should be called `defaultmap`,  "dictionary" is too verbose. And in some existing Lua code we already mention "map-like" or "list-like" tables, so "dictionary" is new, unnecessary jargon.

More generally I don't see `EmptyDictionary` features being used anywhere, so it is like a hidden "feature" that will have surprising behavior. We are introducing lots of new concepts/classes in this PR that should be done in separate PRs (or never). We should use "plain Lua" as much as possible, we should not introduce lots of new classes and object hierarchy.

Reply:
I deleted it because it was in the original PR and probably not used now.